hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/snapshot/SnapshotDiffManager.java

```diff
@@ -131,80 +171,124 @@ private DifferSnapshotInfo getDSIFromSI(SnapshotInfo snapshotInfo,
         getTablePrefixes(snapshotOMMM, volumeName, bucketName));
   }
 
-  @SuppressWarnings("checkstyle:methodlength")
+  @SuppressWarnings("parameternumber")
   public SnapshotDiffReport getSnapshotDiffReport(final String volume,
                                                   final String bucket,
                                                   final OmSnapshot fromSnapshot,
                                                   final OmSnapshot toSnapshot,
                                                   final SnapshotInfo fsInfo,
-                                                  final SnapshotInfo tsInfo)
+                                                  final SnapshotInfo tsInfo,
+                                                  final int index,
+                                                  final int pageSize)
       throws IOException, RocksDBException {
+    String diffJobKey = fsInfo.getSnapshotID() + DELIMITER +
+        tsInfo.getSnapshotID();
+
+    Pair<String, Boolean> jobIdToJobExist = getOrCreateJobId(diffJobKey);
+    String jobId = jobIdToJobExist.getLeft();
+    boolean jobExist = jobIdToJobExist.getRight();
+
+    // If snapshot diff doesn't exist, we generate the diff report first
+    // and add it to the table for future requests.
+    // This needs to be updated to queuing and job status base.
+    if (!jobExist) {
+      generateSnapshotDiffReport(jobId, volume, bucket, fromSnapshot,
+          toSnapshot, fsInfo, tsInfo);
```

Comment:
The diff is a "read" operation from client's perspective, while we are actually writing data in the background. So this request would presumably be served by one of the threads in OM thread pool. And it isn't serialized by Ratis or anything else.

We might need to have a synchronization mechanism here. Otherwise, two requests with the same from/to snapshot could reach OM at the same time and potentially both could reach this line.

Reply:
`getOrCreateJobId()` doesn't look very atomic to me.

Essentially two requests (two separate threads in OM request handler thread pool) can still interleave and reach `getOrCreateJobId` at around the same time. Where they can both get `jobExist == false` and proceed to call `generateSnapshotDiffReport()`.

Reply:
We need to do the table initialization inside `getOrCreateJobId()` when the entry didn't exist.

Reply:
Thanks @smengcl for point it out. Good catch.